src/Features/VisualBasic/Portable/CodeFixes/RemoveUnnecessaryCast/RemoveUnnecessaryCastCodeFixProvider.vb

```diff
@@ -70,117 +44,116 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeFixes.RemoveUnnecessaryCast
+
+            ' Keep track of the all the casts we want to fix up.  We'll fix them up at the end
+            ' after we've done all other manipulation.
+            Dim trackedRoot = editor.OriginalRoot.TrackNodes(originalCastNodes)
+            Dim trackedDocument = document.WithSyntaxRoot(trackedRoot)
+
+            ' Now, go and expand all the containing statements of the nodes we want to edit.
+            ' This is necessary to ensure that the code remains parseable and preserves semantics.
+            Dim expandedRoot = Await ExpandSurroundingStatementsAsync(trackedDocument, originalCastNodes, cancellationToken).ConfigureAwait(False)
+            Dim expandedDocument = document.WithSyntaxRoot(expandedRoot)
+
+            Dim removedRoot = Await RemoveCasts(
+                expandedDocument, originalCastNodes, cancellationToken).ConfigureAwait(False)
+
+            editor.ReplaceNode(editor.OriginalRoot, removedRoot)
+        End Function
 
-            updatedDocument = Await updatedDocument.ReplaceNodeAsync(parentStatement, explicitParentStatement, cancellationToken).ConfigureAwait(False)
+        Private Async Function RemoveCasts(
+                document As Document, originalCastNodes As ImmutableArray(Of ExpressionSyntax),
+                cancellationToken As CancellationToken) As Task(Of SyntaxNode)
 
-            ' Next, make the statement after the parenting statement of the expression semantically explicit.
-            parentStatement = Await FindNodeWithAnnotationAsync(Of StatementSyntax)(s_statementAnnotation, updatedDocument, cancellationToken).ConfigureAwait(False)
-            Dim nextStatement = parentStatement.GetNextStatement()
-            If nextStatement IsNot Nothing Then
-                Dim explicitNextStatement = Await Simplifier.ExpandAsync(nextStatement, updatedDocument, cancellationToken:=cancellationToken).ConfigureAwait(False)
-                updatedDocument = Await updatedDocument.ReplaceNodeAsync(nextStatement, explicitNextStatement, cancellationToken).ConfigureAwait(False)
-            End If
+            Dim root = Await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(False)
 
-            updatedDocument = Await RewriteCoreAsync(updatedDocument, expression, cancellationToken).ConfigureAwait(False)
+            ' Now, find the cast nodes again in the expanded document
+            Dim currentCastNodes = root.GetCurrentNodes(originalCastNodes)
 
-            ' Remove added _expressionAnnotation and _statementAnnotation.
-            updatedDocument = Await RemoveNodesAndTokensWithAnnotationAsync(s_expressionAnnotation, updatedDocument, cancellationToken).ConfigureAwait(False)
-            updatedDocument = Await RemoveNodesAndTokensWithAnnotationAsync(s_statementAnnotation, updatedDocument, cancellationToken).ConfigureAwait(False)
+            Dim innerEditor = New SyntaxEditor(root, document.Project.Solution.Workspace)
```

Comment:
why need inner editor?

Reply:
because we need to do things in two phases.  The first phase complexifies, the second removes casts *within* that complexified code. We need the cast removal code to see the complexified code so it does the right cast removal stuff.  Otherwise, it will just remove the casts fro mthe original code, which doesn't necessarily produce the right stuff in VB cases.

Reply:
I understand that part which require complexity. but was wondering whey not keep using same editor given which tracks changes rather than using new inner editor and replace whole root?

Reply:
Ah, i see the question.  The reason for this is that the old editor belongs to the old document and has a root that matches that old document.  However, we are making these edits a *new* document.  i.e. 'docuemnt' in this method is hte document *after* the complexification has happened.  we can't use the old editor as it doesn't 'belong' to that document.  For example, these 'currentCastNodes' have a different root than the original SyntaxEditor's root.  This will break badly if you ask a syntax editor to make changes to nodes that aren't in its root.

Reply:
so editor.Replace(originRoot, root) and then using that editor doesn't work? anyway, I am fine having more editor, just trying to understand how much support editor provide on manipulating tree.

Reply:
> so editor.Replace(originRoot, root) and then using that editor doesn't work? 

Correct.  All that step is doing is saying "When you do the translation and hit this node, replace it with this other node".  It in no way changes the tree the editor is pointing to.  Basically SyntaxEditor is one-shot.  You set up your operations and you get a new tree.  If you want multiple passes, you use multiple editors.  That's how it's always worked.  

>  just trying to understand how much support editor provide on manipulating tree.

Basically, it's list of opts to apply to a tree at once.  Then you get the new tree from that.

This is similar to how editing with text snapshots work.  You can specify all the edits you want to make.  Then you get the new snapshot, then you can specify all the ones you want to apply to *that*.

In other works, you provide an array of operations.  Not an array of arrays of operations :)  SyntaxEditor doesn't move forward arbitrarily in time.  It is for moving forward one step in time.

--

I would not be opposed to some larger abstraction helping out here, but the nested editor use case is well enough understood (and utilized in several places in our code) that i don't think it would buy much value.